### ISSUE TITLE
watched files are not being uploaded automatically

### ISSUE THREAD

**Issue by Jany-M** [yongkangchen/remote-sync#278]
Hello!
Using `remote-sync 4.1.0` and `Atom 1.7.3`

I added some files to be watched from the GUI panel and I get the notification that they are being watched. However, automatic upload doesnt work upon file change.

If I manually click on Monitor File (the old way), I can see an M icon (doesnt happen with above case though) next to the file in tree view, the notification that file is being watched, but again the file is not uploaded automatically upon change.

Basically, monitoring is now broken, whatever way I try, even the old way.

**Comment by elistone** [comment]
@Jany-M could you give me some details about what OS you are using?
After added files to be monitored have you restarted atom?
Are you sure you are adding the correct path to be automatically followed? A good way to check this is to go to the file you are trying to auto watch: `right click` > `Copy Project Path` remembering to add the beginning slash.

**Comment by Jany-M** [comment]
Win10 64bit.
I have restarted it yes.

I have added them to the config GUI like this:
`/assets/css/responsive.css, /assets/css/extended.css`
and I saw the blu notification:
`remote-sync: Currently watching:`  _the 2 files_

In the `json` they show like this:
`"watch": [
    "/assets/css/responsive.css",
    "/assets/css/extended.css"
  ]`

I dont see any M icon next to those files though.
Only if I manually right click on the file, in tree view, and then click on Monitor File, then the M icon shows up and I get the notification of file being watched.

.css files dont get uploaded automatically anymore though.
Not if I put them in the watch config, nor if I just monitor them manually (the whole monitoring is basically just broken now).

**Comment by elistone** [comment]
Rather strange I am running same OS and do not have the issue, the fact you are seeing the `remote-sync currently watching` implies that when starting up, it is 'watching' the files as checks are made beforehand that the file is actually found.

As for the M not being shown this is more of a style thing (which could be an issue with finding the file or not having time to apply the effect) but I don't class it as a major issue at the moment.

As for a few tests, can you one open atom and using <kbd>ctrl</kbd>+<kbd>P</kbd> look for `Remote Sync: Monitor Files List` this should output the files that you are currently monitoring (just to check they are still in the list once loading is complete)

After that test, could you remove all items from the watch list restart atom & just try doing the manually right click method to help try and narrow it down for me, as I cannot replicate your issue.

Thanks

**Comment by Jany-M** [comment]
After some testing:

- removing the watched files from the config panel, restarting Atom then manually monitoring the files works (watched file is uploaded automatically)
- if then i open the command palette and click on `Remote Sync: Monitor Files List`, in the notice I can still see the old files (even though they are not in the config anymore and in the file it says `"watch": []`). If I restart Atom, the old files dont know up anymore in the List
- if I manually watch a file, then check the List, it's empty (no notice but watched file gets uploaded automatically)

So basically, the old manual way works but **only** if theres no files being watched in the config file.

I use the [keyboard-localization](https://atom.io/packages/keyboard-localization) package (it_IT), if that's of any use, but I tested the above with it disabled.

**Comment by enoversum** [comment]
If I may chime in, I just experience the same on Windows 10 x64. After learning how to put the slashes, Atom finally finds the right file to watch after a restart of the app, the fat M icon appears next to the css file, but it is not being uploaded when changing the .scss file. And I have checked the css file if the change actually applied - yes, the file has been changed, but is not being watched even though it says so :/ .

**Comment by Jany-M** [comment]
had to downgrade to 4.0.1. Hopefully this watch issue is resolved in next version.

One more thing btw: in 4.0.1, when I manually watch for a file, in the popup notice it shows the full path to file. While in 4.1.0 it only shows the file name and extension. Maybe it's nothing, thought I'd share.

@yongkangchen @elistone

**Comment by elistone** [comment]
@Jany-M @enoversum I am struggling to replicate this issue on my machine, and will need access to one with the issue to solve it, you are welcome to look into the issue your self and put in a pull-request.

**Comment by Jany-M** [comment]
@elistone I can try, but do you have any pointers on how to debug this, considering I've never debugged an Atom package nor know .coffee? Maybe files/lines where I could start or steps I should watch (how?) closely?

**Comment by elistone** [comment]
@Jany-M to console log the full pages being watched you can do the following:

**Open the installed package**

- Open settings
- Go to packages
- Find `remote-sync`
- Click to view more
- Click `view code`

**Locate the watching methods** 

Found in `lib/RemoteSync.coffee` - the monitor files methods are found from line **115 - 220**

The method you are most likely looking for is the `mointorFile` method found on line **132** this is where the files are checked to see if they can be watched, if we want to enable toggle and display notices. I would start by console logging the `dirPath` that enters this method and go from there.

Example:

```coffee
monitorFile: (dirPath, toggle = true, notifications = true)->
     console.log "dirPath",dirPath
```

to view the console log you will need to open up the inspector inside atom which can done with the key-bindings <kbd>ctrl</kbd> + <kbd>alt</kbd> + <kbd>I</kbd>

**Important** you will need to reload atom after to see changes take effect. (<kbd>ctrl</kbd> + <kbd>alt</kbd> + <kbd>R</kbd>)

By the way coffeescript is very similar to javascript apart from using indentations, I would recommend using something like [http://js2.coffee/](http://js2.coffee/) to help work out what is going on.

**Comment by Jany-M** [comment]
Thank you @elistone!  I'll try and see what happens.
If I change `lib/RemoteSync.coffee`, do I need to reload Atom each time?

Question... any reason why, both after uninstalling and also after and emptying `.atom/storage`,  `.atom/.node-gyp`  `.atom/compile-cache` ... the `remote-sync` the package looks kinda greyed out in the list (can be installed without issues, I'm just curious)?

![17-05-2016-09-39-39](https://cloud.githubusercontent.com/assets/4365842/15337207/e8266878-1c7a-11e6-939d-642807ead2a0.png)

**Comment by Jany-M** [comment]
@elistone I console logged the paths for each relevant method.

`monitorFile dirPath` `W:\Dropbox\__atom__\project1/assets/css/style.css` (wrong slashes)
`for file in MonitoredFiles file_name replace 1` `W:\\Dropbox\__atom__\project1/assets/css/style.css` (wrong slashes)
`for file in MonitoredFiles file_name replace 2` W:\\Dropbox\\__atom__\\project1/assets/css/style.css (wrong slashes + double slashes)
`setupAutoFileWatch fullpath` `W:\Dropbox\__atom__\project1/assets/css/style.css` (wrong slashes)
`setupAutoFileWatch filesName` `/assets/css/style.css` (wrong slashes)
`initAutoFileWatch projectPath` `W:\Dropbox\__atom__\project1` (correct)

Anything you want me to try here? is this useful to you?

**Comment by elistone** [comment]
@Jany-M Thank you very much! this is very useful, Im busy over the next few days but I can look at it more closely soon. Without checking I'm guessing the issue is to do with lines __166-167__
```
file_name = file.replace(/(['"])/g, "\\$1");
file_name = file.replace(/\\/g, '\\\\');
```
These lines try and escape the string e.g. escape quotes & quadruple slashes for windows to apply the styles but should not be passed to the monitored file function.

or __203__

```
file = dirPath.split('\\').pop().split('/').pop()
```
This function I don't think is an issue just thought I put it out there this is meant to get the file name by popping everything after the last slash.

or __218__

```
fullpath = projectPath + filesName.replace /^\s+|\s+$/g, ""
```

Could be this guy? This function gets the project path and filename and tries to build a url to the file for it to be monitored. 

I think overall there should be a new function created to take care of parsing strings correctly, feel free to look into it, if not I shall try and do it sometime this weekend.

Thanks again for your help, its been great so far!

**Comment by Jany-M** [comment]
@elistone Great! Glad I could be of help.
Considering the wrong slashes only occur with the watchlist files, but not the manual ones, I think the issue isnt so much the replacements themeselves (I thought so too at first), but the monitor watch list file-path management part/functions. 
I'll try to play around a little, but you'll probably have more success ;)

**Comment by elistone** [comment]
@Jany-M great, shall look at in a bit & pull it, thanks again.

**Comment by Jany-M** [comment]
@elistone should we close this issue now or is there anything else to test out?

**Comment by elistone** [comment]
@Jany-M It can be closed, I was just waiting till the patch was pushed.

**Comment by Jany-M** [comment]
@elistone once I updated the package with the latest release and bugfix, I still had to empty Atom's cache/storage in order for the auto-upload to work as expected. Just thought I'd let you know.

**Comment by Jany-M** [comment]
@elistone and it stopped working again :( paths seem right though, so Im not sure where to look now.
It might have stopped working after updating to `Atom 1.8.0`, but I can't be sure.
Is there any chance of a potential conflict with other packages? or is that not possible?
I don't see any errors in the console that could potentially affect this.

**Comment by enoversum** [comment]
@Jany-M It works beautifully with the latest beta on my computers :wink: . So at least it's not broken for everybody.

**Comment by elistone** [comment]
@Jany-M It seems to be broken on my machine too, there have not been any updates to `remote-sync` so it is possible the latest version of Atom has broken it.

**Comment by elistone** [comment]
@Jany-M I found the source of my issue, it was that my root folder had brackets in the name e.g. `folder - details here (something)` the `( )` around the something were creating some kind of error with the watcher [chokidar](https://github.com/paulmillr/chokidar) maybe activating the globing or something.

Could you check if you have strange charters in any of the folders/files to the file you wish to monitor.

**Comment by Jany-M** [comment]
@elistone nope, nothing of the sort unfortunately. It doesnt work on all open projects I have (about 7), and besides a few ones that have a `_` in the folder name, the others are all one work folder names.
Any advice?

**Comment by elistone** [comment]
@Jany-M Sadly nothing I can really do to help in this case as we have not rolled out anything & seems to be local to your machine, you're welcome to look into it yourself.

**Comment by enoversum** [comment]
@Jany-M I had something similar with SASS files not being compiled in Sublime when *any* part of the folder structure had special characters in it. I remember it being ö, ä, ü as well as brackets of any sort, in my case [] . So it failed, even if it was something like this E:\[Folder with brackets]\Projects\My Project . Maybe it is something like that?

**Comment by oueryemchi** [comment]
This is a real life struggle for me :( had Remote-FTP addon working, now busted after the update..tried many and this one seemed the best and now not watching my compiled css, guys is there is any real fix to this? thanks much.
